### PR TITLE
Improve basicAutogradNotImplementedFallback + new tests

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #105592
* #105660
* __->__ #105591
* #105587

This PR:
- removes some reference count bumps (to potentially improve overhead)
- adds some tests for undefined gradients

### CODE REVIEW

File: test/autograd/test_fallback.py [hunk 1]
```diff
@@ -315,6 +315,45 @@ def test_post_autograd_returns_leaf(self, mode):
...
+        with autograd_fallback_mode(mode):
+            lib = self.get_lib()
+            lib.define("foo(Tensor a, Tensor b) -> (Tensor, Tensor)")
+            op = self.get_op("foo")
+
+            def foo_impl(a, b):
+                return (
+                    torch._C._functions.UndefinedGrad()(a),
+                    torch._C._functions.UndefinedGrad()(b),
+                )
```
Comment (soulitzer) [inline test/autograd/test_fallback.py:347]: Are you trying to test the case where your outputs requires_grad=True?  In that case your hooks are registered on the outputs of this function, so they should fire before this happens?

**Review by soulitzer — APPROVED** [review]
Cool optimization! 

For the test, maybe call UndefinedGrad after the op so the undefined grads flow through the hook? Or maybe I misunderstood what it is trying to test.

File: test/autograd/test_fallback.py (hunk 1 above)
Comment (zou3519) [reply test/autograd/test_fallback.py:347]: Good point
